File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.js
function solution(maps) {
    const answer = [],
        [row, col] = [maps.length, maps[0].length],
        [dirX, dirY] = [[0, 0, -1, 1], [-1, 1, 0, 0]],
        visited = Array.from(Array(row),() => new Array(col).fill(false));

    const dfs = (x, y) => {
        let food = maps[x][y] * 1;
        visited[x][y] = true;

        for (let i = 0; i < 4; ++i) {
            const [destX, destY] = [x + dirX[i], y + dirY[i]];
            if (destX < 0 || destX >= row || destY < 0 || destY >= col) continue;
            if (isNaN(maps[destX][destY]) || visited[destX][destY]) continue;
             
            food += dfs(destX, destY);
        }
        return food;
    };

    maps.forEach((map, x) => {
        [...map].forEach((el, y) => {
            if (isNaN(el) || visited[x][y]) return;
            
            answer.push(dfs(x, y));
        });
    });
    return answer.length ? answer.sort((a, b) => a - b) : [-1];
}

File: 프로그래머스/unrated/154540. 무인도 여행/무인도 여행.cpp
#include <string>
#include <vector>
#include <algorithm>

using namespace std;

int dfs(int x, int y, vector<vector<bool>>& visited, const vector<pair<int, int>>& dir, const vector<string>& maps)
{
    int value = maps[x][y] - '0';
    visited[x][y] = true;

    for (const auto& d : dir)
    {
        int tx = x + d.first,
            ty = y + d.second;

        if (tx < 0 || tx >= maps.size() || ty < 0 || ty >= maps[0].length()) continue;
        if (maps[tx][ty] == 'X' || visited[tx][ty]) continue;

        value += dfs(tx, ty, visited, dir, maps);
    }
    return value;
}

vector<int> solution(vector<string> maps) {
    vector<int> answer;
    size_t row = maps.size(), col = maps[0].length();
    vector<vector<bool>> visited(row, vector<bool>(col, false));
    const vector<pair<int, int>> dir = { {0, -1}, {0, 1}, {1, 0}, {-1, 0} };
    
    for (size_t x = 0; x < row; ++x)
    {
        for (size_t y = 0; y < col; ++y)
        {
            if (maps[x][y] == 'X' || visited[x][y]) continue;

            answer.emplace_back(dfs(x, y, visited, dir, maps));
        }
    }

    if (answer.empty()) answer.emplace_back(-1);
    else ::stable_sort(answer.begin(), answer.end());

    return answer;
}